core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AbstractMaterializeTableScan.java

```diff
@@ -0,0 +1,217 @@
+        this.pageSourceManager = requireNonNull(pageSourceManager, "pageSourceManager is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.currentNode = nodeManager.getCurrentNode().getHostAndPort();
+    }
+
+    @Override
+    public final boolean isEnabled(Session session)
+    {
+        return isMaterializeTableEnabled(session) &&
+                getMaterializeTableMaxEstimatedRowCount(session) > 0 &&
+                getMaterializeTableMaxActualRowCount(session) > 0 &&
+                getMaterializeTableTimeout(session).toMillis() > 0;
+    }
+
+    @Override
+    public final Result apply(T node, Captures captures, Context context)
+    {
+        int maxRows = getMaterializeTableMaxEstimatedRowCount(context.getSession());
+        PlanNodeStatsEstimate stats = context.getStatsProvider().getStats(node);
+        if (stats.getOutputRowCount() > maxRows) {
+            return Result.empty();
+        }
+
+        return apply(node, captures, context.getSession());
+    }
+
+    protected abstract Result apply(T node, Captures captures, Session session);
+
+    protected Optional<List<Expression>> materializeTable(Session session, TableScanNode tableScan, Constraint constraint)
+    {
+        TableHandle table = tableScan.getTable();
+        List<ColumnHandle> columns = tableScan.getOutputSymbols().stream()
+                .map(symbol -> tableScan.getAssignments().get(symbol))
+                .toList();
+        List<Type> types = tableScan.getOutputSymbols().stream()
+                .map(Symbol::type)
+                .toList();
+
+        Future<Optional<List<Expression>>> future = executor.submit(() -> {
+            try (var _ = new SetThreadName("MaterializeTableScan-%s", table.catalogHandle())) {
+                return doMaterializeTable(session, table, columns, types, constraint);
+            }
+        });
+
+        Duration timeout = getMaterializeTableTimeout(session);
+        try {
+            return future.get(timeout.toMillis(), MILLISECONDS);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            future.cancel(true);
+            throw new RuntimeException("interrupted", e);
+        }
+        catch (ExecutionException e) {
+            if (e.getCause() instanceof TableTooLargeException) {
+                log.warn("Table too large to materialize: %s", table);
+            }
+            else {
+                log.warn(e.getCause(), "Failed to materialize table: %s", table);
+            }
+        }
+        catch (TimeoutException e) {
+            log.warn("Timed out while materializing table: %s", table);
+            future.cancel(true);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<List<Expression>> doMaterializeTable(Session session, TableHandle table, List<ColumnHandle> columns, List<Type> types, Constraint constraint)
```

Comment:
Wouldn't we want to run this again after a filter is pushed? A table might be too large initially, but comes eligible when when more information is available.

Reply:
> A table might be too large initially, but comes eligible when when more information is available.

But that would mean multiple few secs delays to bail-out, right? Splits would be enumerated multiple times. JDBC queries would also be triggered on remote system multiple times.

I think it should have just one shot after all eligible pushdowns, BUT then we also want materialized predicate to be used for further pushdowns. It's chicken and egg problem.

My guess it should just try once and predicates derived from materialized tables should be applied on `ValuesNodes`. It would solve 90% of small table cases this way.